### PR TITLE
Documentation for the kernel API

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

This PR provides the documentation for the big API changes and extension provided by #46339

### CODE REVIEW

File: docs/source/en/kernel_doc/writing_kernels.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+2. **Module fusion** — the kernel replaces multiple adjacent modules with a single fused implementation.
+
+For basic kernels (stateless `forward` replacements with no parameter changes), see the [kernels library](https://github.com/huggingface/kernels) documentation.
+
+## Two-class pattern
+
+Any kernel that carries its own parameters follows a two-class pattern.
+
+- `KernelName`: contains only the `forward` pass. The `kernels` library uses this class to kernelize the model. It must be stateless as the library does not allow stateful kernel classes.
+- `KernelNameLayout`: an `nn.Module` that holds the parameters and monkey-patches the original module before the checkpoint is loaded. At runtime, `kernelize` replaces its `forward` with `KernelName`'s `forward`. You do not need to define `forward`, Transformers injects one automatically with the same signature as `KernelName.forward`.
```
Comment (stevhliu) [inline docs/source/en/kernel_doc/writing_kernels.md]:
```suggestion
- `KernelNameLayout`: an `nn.Module` that holds the parameters and monkey-patches the original module before the checkpoint is loaded. At runtime, `kernelize` replaces its `forward` with the `forward` from `KernelName`. You don't need to define `forward`. Transformers injects one automatically with the same signature as `KernelName.forward`.
```

File: docs/source/en/kernel_doc/writing_kernels.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+1. **Parameter transformation** — the kernel expects weights in a different layout than the original model (for example, renamed or merged parameters).
+2. **Module fusion** — the kernel replaces multiple adjacent modules with a single fused implementation.
+
+For basic kernels (stateless `forward` replacements with no parameter changes), see the [kernels library](https://github.com/huggingface/kernels) documentation.
+
+## Two-class pattern
+
+Any kernel that carries its own parameters follows a two-class pattern.
+
+- `KernelName`: contains only the `forward` pass. The `kernels` library uses this class to kernelize the model. It must be stateless as the library does not allow stateful kernel classes.
```
Comment (stevhliu) [inline docs/source/en/kernel_doc/writing_kernels.md]:
```suggestion
- `KernelName`: contains only the `forward` pass. The `kernels` library uses this class to kernelize the model because it doesn't allow stateful kernel classes.
```

File: docs/source/en/kernel_doc/writing_kernels.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+rendered properly in your Markdown viewer.
+
+-->
+
+# Writing kernels
+
+This guide explains how to write kernels that go beyond a stateless `forward` replacement. It covers two capabilities the extended `KernelConfig` API supports:
+
+1. **Parameter transformation** — the kernel expects weights in a different layout than the original model (for example, renamed or merged parameters).
+2. **Module fusion** — the kernel replaces multiple adjacent modules with a single fused implementation.
```
Comment (stevhliu) [inline docs/source/en/kernel_doc/writing_kernels.md]:
```suggestion
1. Parameter transformation — the kernel expects weights in a different layout than the original model (for example, renamed or merged parameters).
2. Module fusion — the kernel replaces multiple adjacent modules with a single fused implementation.
```

File: docs/source/en/kernel_doc/writing_kernels.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+For basic kernels (stateless `forward` replacements with no parameter changes), see the [kernels library](https://github.com/huggingface/kernels) documentation.
```
Comment (stevhliu) [inline docs/source/en/kernel_doc/writing_kernels.md]:
```suggestion
For basic kernels (stateless `forward` replacements with no parameter changes), see the [kernels](https://github.com/huggingface/kernels) library documentation.
```

File: docs/source/en/kernel_doc/writing_kernels.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+## Two-class pattern
+
+Any kernel that carries its own parameters follows a two-class pattern.
+
+- `KernelName`: contains only the `forward` pass. The `kernels` library uses this class to kernelize the model. It must be stateless as the library does not allow stateful kernel classes.
+- `KernelNameLayout`: an `nn.Module` that holds the parameters and monkey-patches the original module before the checkpoint is loaded. At runtime, `kernelize` replaces its `forward` with `KernelName`'s `forward`. You do not need to define `forward`, Transformers injects one automatically with the same signature as `KernelName.forward`.
+
+**The naming convention is strict: the layout class must be named `{KernelName}Layout` and defined in the same module as `KernelName`. Transformers discovers it automatically.**
```
Comment (stevhliu) [inline docs/source/en/kernel_doc/writing_kernels.md]:
```suggestion
> [!IMPORTANT]
> The naming convention is strict. The layout class must be named `{KernelName}Layout` and defined in the same module as `KernelName`.
```

File: docs/source/en/kernel_doc/writing_kernels.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+## Parameter transformation
+
+Use this pattern when the kernel expects weights under different names or in a different shape than the original model checkpoint.
+
+The `KernelNameLayout` class has the **same `__init__` signature as the module it replaces** and declares a `conversion_mapping` class attribute that tells Transformers how to remap checkpoint keys to the new parameter names.
```
Comment (stevhliu) [inline docs/source/en/kernel_doc/writing_kernels.md]:
```suggestion
The `KernelNameLayout` class has the same `__init__` signature as the module it replaces and declares a `conversion_mapping` class attribute that tells Transformers how to remap checkpoint keys to the new parameter names (see [Dynamic weight loading](../weightconverter) for more details).
```

File: docs/source/en/kernel_doc/writing_kernels.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+
+class layers:
+    CustomRMSNorm = CustomRMSNorm
+```
+
+> [!NOTE]
+> The `layers` class is required by the `kernels` library to expose the kernel entry point.
+
+Load this kernel by passing the repo and class name to `KernelConfig`. The key is the **original** module class name from the model; the value points to the `KernelName` class (not the Layout) in the repo.
```
Comment (stevhliu) [inline docs/source/en/kernel_doc/writing_kernels.md]:
```suggestion
Load this kernel by passing the repo and class name to [`KernelConfig`]. The key is the original module class name from the model. The value points to the `KernelName` class (not the `Layout`) in the repo.
```

File: docs/source/en/kernel_doc/writing_kernels.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+When the model loads, Transformers:
+1. Loads `CustomRMSNorm` from the repo and looks for `CustomRMSNormLayout` in the same module.
+2. Monkey-patches every `RMSNorm` in the model with `CustomRMSNormLayout`.
+3. Remaps checkpoint weights using `conversion_mapping` so they load into the new parameter names.
+4. Calls `kernelize`, which replaces `CustomRMSNormLayout.forward` with `CustomRMSNorm.forward`.
+
+## Module fusion
+
+Use this pattern when a kernel replaces multiple adjacent modules with a single fused implementation. Because the fused module combines parameters from several original modules, the `KernelNameLayout.__init__` receives the **instantiated child modules** rather than their constructor arguments.
```
Comment (stevhliu) [inline docs/source/en/kernel_doc/writing_kernels.md]:
```suggestion
Use this pattern when a kernel replaces multiple adjacent modules with a single fused implementation. Because the fused module combines parameters from several original modules, the `KernelNameLayout.__init__` receives the instantiated child modules rather than their constructor arguments.
```

File: docs/source/en/kernel_doc/writing_kernels.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        hidden_states = self.scale * hidden_states.to(input_dtype)
+        gate, up = self.gate_up_proj(hidden_states).chunk(2, dim=-1)
+        return self.down_proj(self.act_fn(gate) * up)
+
+
+class layers:
+    RMSNormMLP = RMSNormMLP
+```
+
+To fuse modules, pass a **tuple of `(class_name, path_pattern)` pairs** as the key in `KernelConfig` instead of a plain string. All patterns must share the same parent module (Transformers fuses the children in that parent). The `*` wildcard matches any single path segment.
```
Comment (stevhliu) [inline docs/source/en/kernel_doc/writing_kernels.md]:
```suggestion
To fuse modules, pass a tuple of `(class_name, path_pattern)` pairs as the key in `KernelConfig` instead of a plain string. All patterns must share the same parent module (Transformers fuses the children in that parent). The `*` wildcard matches any single path segment.
```

**Review by stevhliu — APPROVED** [review]
thanks, this is a very solid set of docs!

File: docs/source/en/kernel_doc/writing_kernels.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+Use this pattern when the kernel expects weights under different names or in a different shape than the original model checkpoint.
+
+The `KernelNameLayout` class has the same `__init__` signature as the module it replaces and declares a `conversion_mapping` class attribute that tells Transformers how to remap checkpoint keys to the new parameter names (see [Dynamic weight loading](../weightconverter) for more details).
+
+```python
+import torch
+import torch.nn as nn
+
+class CustomRMSNormLayout(nn.Module):
+    conversion_mapping = [...]  # rules that remap checkpoint keys to the new parameter names
```
Comment (vasqu) [inline docs/source/en/kernel_doc/writing_kernels.md:48]: Imo we should note that this is transformers specific and kernels doesn't ship this dependency (or even as optional one) so this needs some guarding https://huggingface.co/kernels/AntonV/dummy-rmsnorm-mlp-with-transformations-and-init/blob/v0/build/torch-cuda/layers.py

File: docs/source/en/kernel_doc/writing_kernels.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+class CustomRMSNorm(nn.Module):
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        input_dtype = hidden_states.dtype
+        hidden_states = hidden_states.to(torch.float32)
+        variance = hidden_states.pow(2).mean(-1, keepdim=True)
+        hidden_states = hidden_states * torch.rsqrt(variance + self.variance_epsilon)
+        return self.scale * hidden_states.to(input_dtype)
+
+
+class layers:
```
Comment (vasqu) [inline docs/source/en/kernel_doc/writing_kernels.md:65]: imo this is dirty, we should advice to use the proper structure of layers.py

File: docs/source/en/kernel_doc/writing_kernels.md [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+
+> [!NOTE]
+> The `layers` class is required by the `kernels` library to expose the kernel entry point.
+
+Load this kernel by passing the repo and class name to [`KernelConfig`]. The key is the original module class name from the model. The value points to the `KernelName` class (not the `Layout`) in the repo.
+
+```python
+from transformers import AutoModelForCausalLM, KernelConfig
+
+kernel_config = KernelConfig({"RMSNorm": "owner/my-kernel:CustomRMSNorm"})
```
Comment (vasqu) [inline docs/source/en/kernel_doc/writing_kernels.md:77]: this wont work because it's a remote kernel, you would need to pass some way of trusting it
```suggestion
kernel_config = KernelConfig({"RMSNorm": ("owner/my-kernel:CustomRMSNorm", {"trust_remote_code": True})})
```
Also note that it needs a version or revision; if there is not the default version 1 then it will also crash

**Review by vasqu — COMMENTED** [review]
Some comments from my side because there is some outdated info imo. This comes with the bump to the newest kernels as there are stricter requirements and model type repos are not allowed anymore

I used torch-no-arch builders to produce the new kernels

**Comment by michaelbenayoun** [comment]
I addressed your comments here: #46889
